### PR TITLE
MapSketchImporter for d2k, ra and cnc mods.

### PR DESCRIPTION
**Pull request by CH4Code**
**Short**: This puts a new button "Import Sketch" into the map editor menue which interactively allows to use bitmap (.bmp) picture files for pre-filling a map with default terrain on d2k, ra and cnc.

**Illustrated**:
converts this
mapinter
into this (without further editing)
interrior

The color mapping is interactive for 13 predefined color-ranges:
color-remap

Works for all three standard mods and all terrain sets. Can also generate resources and trees.
A [bitmap like this](https://user-images.githubusercontent.com/16046997/34074104-88f3e57e-e2a8-11e7-85ff-b73783732b97.png) (this is png because of upload reasons) can be used with different color remaps to generate three different maps in the three mods ([cnc-map](https://user-images.githubusercontent.com/16046997/34074112-a62e93d2-e2a8-11e7-947c-b45832f6e0ca.png),[d2k-map](https://user-images.githubusercontent.com/16046997/34074113-a6483422-e2a8-11e7-9f9d-74d83abf378b.png), [ ra-map](https://user-images.githubusercontent.com/16046997/34074114-a6647880-e2a8-11e7-993f-41f78d3b554c.png) ).

**Verbose**:
I made this to increase map making speed, since you often start with a sketch anyway. This way you can use this sketch directly to generate a map. Mirroring maps is very easy in any paint program and hard in the ora editor. The interfaces between terrain (like cliffs) still have to be placed manually to look appealing, but the rough layout is finished with the import.
The full color space is utilized, but the boundaries are somewhat arbitrary and could be refined further. I went for maximal human-subjective distinguishability. I use HSB color space with Hue beeing the main distinguishing factor.
Which terrain is default and which is selectable for each tileset (with the dropdowns) is hardcoded and maybe better moved to yaml files, but I just wanted to get this out for discussion..

Other improvements which could be made would be
* in-menu map preview with highlighting of the color areas if the mouse cursor hovers over a selection row.
* automatic place-and-fit of cliff and other connected tiles.
* color selection dropdowns like player color dropdown, however that is a bit tricky since the borders of "adjacent colors" will move.
* support png pictures

Please consider this as a first version wich could improve over time.. 
Also my first c# code commit here, so please go easy on me :-D

### CODE REVIEW

**Comment by CH4Code** [comment]
I broke it somehow right before uploading, and dont know where the error is.  Pls wait until it is fixed.   -.-

**Comment by MustaphaTR** [comment]
I'm not sure if shortening Light and Dark is necessary, it looks like  they would fit anyway.

**Comment by CH4Code** [comment]
Fixed!

@MustaphaTR If you mean white( brightness>90%) and black (brightness<10%) beeing restricted, that is because the remaining 80% of brightness levels are used for colors. all colors become white at high brightness, all colors become black at low brightness.

**Comment by MustaphaTR** [comment]
No i mean the text on the UI. `Light` and `Dark` instead of `Lt. ` and `Drk. `.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/MapSketchImporterLogic.cs [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+		private enum ColorIndex { White, Black, Gray, Red, Orange, Yellow, Green, Teal, LightBlue, DarkBlue, Purple, Pink, Magenta }
+
+		class DropDownOption
+		{
+			public string Title;
+			public Func<bool> IsSelected;
+			public Action OnClick;
+		}
+
+		static Dictionary<int, string> defaultFillerTilesd2k = new Dictionary<int, string>
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/MapSketchImporterLogic.cs]: Sorry, but you can't hard-code mod specific (d2k) stuff in OpenRA.Mods.**Common**.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/MapSketchImporterLogic.cs [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+
+			if (modData.Manifest.Id == "ra")
+			{
+				activeModDefaultFillers = modDefaultFillersRA;
+				activeModFillers = modFillersRA;
+				activeResourceFillers = fillerResourcesRA;
+				activeDefaultResourceFillers = defaultFillerResourcesRA;
+			}
+
+			if (modData.Manifest.Id == "cnc")
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/MapSketchImporterLogic.cs]: Same here.

**Review by Mailaender — CHANGES_REQUESTED** [review]
I really like this idea in general. However the GUI integration is problematic and maybe even overkill. How about making this a command for `OpenRA.Utility.exe` with mod-specific configuration files instead?

**Comment by pchote** [comment]
Re GUI integration: @abcdefg30 recently suggested resurrecting the abandoned TilesetBuilder on top of the Mod SDK.  Building external tools using the SDK is a great idea that hadn't occurred to me earlier: it lets us build tools using our cross-platform (albeit horrible to develop against) widget framework without adding the maintenance burden of continuously maintaining them upstream or the complexity burden for normal players.

**Comment by CH4Code** [comment]
(A) I'm kind of sceptical about the utility command solution, because it would require oblong input arguments or reconfiguring a config file each time before conversion. Since it can't be interactive, the result will only be visible after restarting openra, making trial and error a painfull process. And it would be hardly visible/hard to use for the average joe map maker. (How many people even know how to invoke a command line?). The "Import Sketch" button instantly reminds a mapmaker that there is this option, too.

(B) One of the advantages of including it in the main project (for me) is that openra is a trusted and tested environment, so people need not think about downloading and installing something from an untrusted source (like me). And I dont need to worry too much about licensing, warranties or of linux/mac compatability... So people would actually use it ^^

I think I'll give B a shot anyways and try to rip everything I don't need from the SDK to make it an external tool for map sketch importing then. 

Any other comments?

**Pull request closed without merging**